docs: extend timer example with buttons (#195)
aica-technology/api: docs/docs/getting-started/04-examples/02-editor-example.md

You can also use the graph control buttons on the bottom left to zoom and fit the view. The mini-map on the bottom right
can also be used to navigate around the graph.

> Component parameters and settings are also available through the AICA Studio. Click on the gear icon of the first Timer
> component to open its settings page and learn about auto-lifecycle events and component parameters in the sections
> below.

Reviewer: ```suggestion
## Editing component settings

Component parameters and settings can be edited directly from the graph view. Click on the gear icon of the first Timer
component to open its settings panel. The settings panel can be closed with the X or by clicking anywhere in the graph. 
```

![timer settings](./assets/timer-settings.png)

> ## Auto-lifecycle events

Reviewer: ```suggestion
### Auto-lifecycle events
```

Because the timer components are lifecycle components, they only count down the time when they are active. By default,
the "Load" or "Transition" events start the lifecycle component in an unconfigured state. The "auto-configure" and
"auto-activate" toggle switches can be used to enable or disable
[auto-lifecycle events](../../concepts/05-building-blocks/03-components.md#auto-lifecycle-events).

When both toggle switches are enabled, the YAML code will show the corresponding events for the componnts:

```yaml
    events:
      transitions:
        on_load:
          lifecycle:
            component: timer
            transition: configure
        on_configure:
          lifecycle:
            component: timer
            transition: activate
```

Disabling one or both auto-lifecycle event switches will also remove the corresponding events from the YAML, while
enabling the switch will regenerate the event in the YAML.

> ## Setting component parameters

Reviewer: ```suggestion
### Setting component parameters
```

> Component parameters can be edited directly in the settings. Try changing the value of the Timeout parameter and look

Reviewer: ```suggestion
Component parameters can be edited directly in the settings panel. Change the value of the Timeout parameter and look
```

different duration before triggering the "Is timed out" predicate.

<!-- TODO: explain behavior with "default" once it's working as intended -->

## Adding and deleting elements

Let's now make the example a bit more interesting by adding more elements to the graph using the sidebar menu; press the
(+) button in the corner of the graph to open the sidebar. At the top, you will see special entries
"Hardware Interface", "Trigger Events Button", "Sequence" and "Condition", for which a dedicated example is included
in later sections. These are followed by a list of components with a name and brief description.

<!-- TODO: link the examples once they exist -->

Trigger Events Buttons are interactive elements that can be used to trigger events through mouse clicks and interact
with the flow of the application. Click on the Trigger Events Button in the sidebar to add such an element to the
application. A new button should appear on the graph and in the code.

The sidebar menu can be closed by clicking anywhere in the graph. Drag the button to change its position in the graph.
This will also update the corresponding Trigger Events Button position in the code editor.

> To delete an element from the graph, press the small menu icon in its top right corner, then click Remove.

Reviewer: ```
Elements can be added to the graph using the sidebar menu; press the (+) button in the top right corner of the graph to
open the sidebar and see a list of available application elements. At the top of the list are the "Hardware Interface",
"Trigger Events Button", "Sequence" and "Condition" nodes. These are followed by a list of all available components from
AICA Core and any additionally installed packages, grouped by package. Clicking on any element in the sidebar will
automatically add it to the graph.

To delete an element from the graph, press the small menu icon in its top right corner, then click Remove. Elements can
also be deleted by selecting them with a click and pressing the Delete keyboard shortcut.

For this example, let's add a "Trigger Events Button" to the timer application by clicking on it in the sidebar. A new
button should appear on the graph and in the code. This is an interactive elements that can be used to trigger events
through mouse clicks and interact with the flow of the application.

The sidebar menu can be closed by the X icon or by clicking anywhere in the graph.

Drag the button to change its position in the graph. This will also update the corresponding Trigger Events Button
position in the YAML code editor.
```

## Creating and deleting event edges

To create a new edge, move the mouse over an event source handle until the cursor changes to a targeting reticule. Then,
click and drag to create a draft edge. Pull it towards a target handle (on the top left of a component) until the cursor
changes again and the draft edge snaps in place, at which point you can let go of the mouse button. If the connection is
valid, it will create a new event edge with the default event (Load). Event source handles are found under transitions
and predicates of a component and look like that:

![event handle](./assets/event-source-handle.png)

:::note

> All components have an icon on their top right that looks like an event source handle but is in fact a condition source
> and creating event edges from those will not work. 

Reviewer: ```suggestion
All components also have a similar source handle icon on the top right. This handle is not an event source and cannot be
connected to target components or hardware with event edges. Instead, it is a condition source used to connect condition
edges to condition and sequence inputs.

Invalid edge connections will automatically be rejected.
```

![event edge](./assets/event-edge.png)

Clicking on an edge also selects it, which is indicated by the increased line thickness. Clicking away from the edge
will deselect it. While the edge is selected, press the delete key to delete the edge. This will also remove the
event from the YAML representation.

To create a new edge, move the mouse over the event source handle next to a component predicate until the cursor changes
to a targeting reticule. Then, click and drag to create a draft edge. Pull it towards a target handle (on the top left
of a component) until the cursor changes again and the draft edge snaps in place, then let go of the mouse button.

If the connection is valid, it will create a new event edge with the default event (Load).

## Adding and deleting components
Add now a second trigger button, create a new event edge to Timer 2 and choose the Activate event type.

Components can be added to the graph using the component sidebar menu; press the (+) button in the corner of the graph
to open the sidebar. At the top, you will see two special entries "Event Trigger" and "Hardware Interface", which will

> be explained in a later example. These are followed by a list of components with a name and brief description.

Reviewer: ```
Create an event edge between the new trigger button and the Timer 2 component. The default event type newly created edge
is always "Load". To change the type of event that should be triggered, click on the event label on the edge to open a
selection menu showing other available event types (for example, Load, Unload, Configure...) and click on the desired
event. This will close the selection menu and update the YAML code accordingly.

For this example, choose the Deactivate event in order to pause the timer on click of the button.

![event edge](./assets/event-edge.png)

Clicking on an edge also selects it, which is indicated by the increased line thickness. Clicking away from the edge
will deselect it. While the edge is selected, press the Delete key to delete the edge. This will also remove the
event from the YAML representation.

Now add a second trigger button, create a new event edge to Timer 2 and choose the Activate event type.
```

> scratch using only the graph editor. Remember to set a timeout value in the component parameters.

Reviewer: ```
Both buttons now still have the default "Trigger Events Button" name, which may become confusing in larger applications.
To rename an element in the graph, click on the small menu icon in its top right corner, then click Rename. Give the two
buttons a new name: "Pause Timer" for the button with the Deactivate event, and "Play Timer" for the button with the
Activate event.

Now, try to run the applications again and use the Pause Timer and Play Timer trigger buttons to deactivate and activate
Timer 2 while it is active. Verify that once deactivated, the elapsed time does not count towards the component timing
out and it only times out once it has been in the active state for a total of 4 seconds. That is, deactivating the
component will freeze the timer until reactivation.

Now that we learned to add buttons and create event edges, try to do the same for components. As an exercise, delete
all elements from the graph and try to recreate this example from scratch using only the graph editor. Search for and
add the timer components from the sidebar, set the auto-lifecycle events, and connect the transition edges. Remember to
set a timeout value in the component parameters.
```

enabling the switch will regenerate the event in the YAML.

## Setting component parameters

Component parameters can be edited directly in the settings panel. Change the value of the Timeout parameter and look
for the corresponding change in the code editor. Then press play on the application and observe that the timer now has a
different duration before triggering the "Is timed out" predicate.

<!-- TODO: explain behavior with "default" once it's working as intended -->

## Adding and deleting elements

Elements can be added to the graph using the sidebar menu; press the (+) button in the top right corner of the graph to
open the sidebar and see a list of available application elements. At the top of the list are the "Hardware Interface",
"Trigger Events Button", "Sequence" and "Condition" nodes. These are followed by a list of all available components from
AICA Core and any additionally installed packages, grouped by package. Clicking on any element in the sidebar will
automatically add it to the graph.

<!-- TODO: link the examples once they exist -->

To delete an element from the graph, press the small menu icon in its top right corner, then click Remove. Elements can

> also be deleted by selecting them with a click and pressing the Delete keyboard shortcut.

Reviewer 2: ```suggestion
also be deleted by selecting them with a click and pressing the Delete key on your keyboard.
```

Author: This is actually a good point because it's not correct...Non-mac keyboards have a backspace and a delete key whereas mac keyboards only have a delete. For non-mac, only the backspace key works

Reviewer 2: huh, yes true. We should then probably mention it in a platform-specific way

Author: Even better, we take care of the keybindings explicitly in the future. For now, I made platform specific tabs

Reviewer 2: nice, thanks! Should we through Windows in too (e.g., Linux/Windows)? :smile: or not just yet

Author: Not just yet :smile:

Click on the gear icon of the first Timer component to open its settings page and learn about auto-lifecycle events and
component parameters in the sections below.

![timer settings](./assets/timer-settings.png)

## Auto-lifecycle events

Because the timer components are lifecycle components, they only count down the time when they are active. By default,
the "Load" or "Transition" events start the lifecycle component in an unconfigured state. The "auto-configure" and
"auto-activate" toggle switches can be used to enable or disable
[auto-lifecycle events](../../concepts/05-building-blocks/03-components.md#auto-lifecycle-events).

When both toggle switches are enabled, the YAML code will show the corresponding events for the componnts:

```yaml
    events:
      transitions:
        on_load:
          lifecycle:
            component: timer
            transition: configure
        on_configure:
          lifecycle:
            component: timer
            transition: activate
```

Disabling one or both auto event switches will also remove the corresponding events from the YAML, while enabling the
switch will regenerate the event in the YAML.

## Setting component parameters

> Component parameters can be edited directly in the settings. Try changing the number of the Timeout parameter to a
> different value in seconds, and look for the corresponding change in the code editor. Then press play on the application
> and observe that the timer now has a different duration before triggering the "Is timed out" predicate.

Reviewer 2: ```suggestion
Component parameters can be edited directly in the settings. Try changing the value of the
Timeout parameter to a different one and look for the corresponding change in the code 
editor. It only takes short moment to make the change and propagate it to the code! Then 
press play on the application and observe that the timer now has a different duration before 
triggering the "Is timed out" predicate.
```

## Adding and deleting elements

Let's now make the example a bit more interesting by adding more elements to the graph using the sidebar menu; press the
(+) button in the corner of the graph to open the sidebar. At the top, you will see special entries

> "Hardware Interface", "Trigger Events Button", "Sequence" and "Condition", which will all have a dedicated example later
> on. These are followed by a list of components with a name and brief description.

Reviewer 2: ```suggestion
"Hardware Interface", "Trigger Events Button", "Sequence" and "Condition", for which a dedicated example is included  in later
sections. These are followed by a list of components with a name and brief description.
```

Trigger Events Buttons are interactive elements that can be used to trigger events through mouse clicks and interact
with the flow of the application. Click on the Trigger Events Button in the sidebar to add such an element to the
application. A new button should appear on the graph and in the code.

> The sidebar menu can be closed by clicking anywhere in the graph. Click and drag on the button to change its position in

Reviewer 2: ```suggestion
The sidebar menu can be closed by clicking anywhere in the graph. Drag on the button to change its position in
```

> the graph. This will also update the associated position in the code editor.

Reviewer 2: ```suggestion
the graph. This will also update the corresponding Trigger Event Button position in the code editor.
```

To delete an element from the graph, press the small menu icon in its top right corner, then click Remove.

## Creating and deleting event edges

The timers are already connected with a Transition event edge. To change the event type on an edge, click on the label.
This will open a selection menu showing other available event types (for example, Load, Unload, Configure...). Choosing
a different event type will close the selection menu and update the YAML code accordingly.

![event edge](./assets/event-edge.png)

Clicking on an edge also selects it, which is indicated by the increased line thickness. Clicking away from the edge
will deselect it. While the edge is selected, press the delete key to delete the edge. This will also remove the
event from the YAML representation.

> To create a new edge, move the mouse over the event source handle until the cursor changes to a targeting reticule.

Reviewer 2: ```suggestion
To create a new edge, move your mouse over the event source handle until the cursor changes to a targeting reticle.
```

perhaps both are valid :thinking:

Author: I think both are :+1: It's not in my vocabulary and I took the Docs version :smile:

Then, click and drag to create a draft edge. Pull it towards a target handle (on the top left of a component) until the

> cursor changes again and the draft edge snaps in place, then let go of the mouse button. If the connection is valid, it

Reviewer 2: ```suggestion
cursor changes again and the draft edge snaps in place, at which point you can let go of the mouse button. If the connection is valid, it
```

will create a new event edge with the default event (Load).

> Follow this procedure to create an event between the new trigger button and the Timer 2 component. Choose the Deactivate

Reviewer 2: ```suggestion
Repeat this procedure to create an event between the new trigger button and the Timer 2 component. Choose the Deactivate
```

event type in order to pause the timer on click of the button. Add a second trigger button, create a new event edge to
Timer 2 and choose the Activate event type.

![trigger buttons](./assets/trigger-buttons.gif)

## Renaming elements in the graph

> Both buttons now still have the default "Trigger Events Button" name and should be renamed. To rename an element in the
> graph, click on the small menu icon in its top right corner, then click Rename. Give the two buttons a new name, for
> example "Pause Timer" for the button with the Deactivate event, and "Play Timer" for the button with the Activate event.

Reviewer 2: ```suggestion
Both buttons now still have the default "Trigger Events Button", which may become confusing in larger applications. To rename an element in the
graph, click on the small menu icon in its top right corner, then click Rename. Give the two buttons a new name, for
example "Pause Timer" for the button with the Deactivate event, and "Start Timer" for the button with the Activate event.
```

> Try to run the applications again now and use the Pause and Play trigger buttons to deactivate and activate the
> Timer 2 while it is active. Verify that once paused, the elapsed time does not count towards the component timing out
> and it only times out once it has been in the active state for a total of 4 seconds.

Reviewer 2: ```suggestion
Now, try to run the applications again and use the Pause and Play trigger buttons to deactivate and activate the
Timer 2 while it is active. Verify that once paused, the elapsed time does not count towards the component timing out
and it only times out once it has been in the active state for a total of 4 seconds. That is, deactivating the component will freeze the timer until reactivation.
```

below.

![timer settings](./assets/timer-settings.png)

## Auto-lifecycle events

Because the timer components are lifecycle components, they only count down the time when they are active. By default,
the "Load" or "Transition" events start the lifecycle component in an unconfigured state. The "auto-configure" and
"auto-activate" toggle switches can be used to enable or disable
[auto-lifecycle events](../../concepts/05-building-blocks/03-components.md#auto-lifecycle-events).

When both toggle switches are enabled, the YAML code will show the corresponding events for the componnts:

```yaml
    events:
      transitions:
        on_load:
          lifecycle:
            component: timer
            transition: configure
        on_configure:
          lifecycle:
            component: timer
            transition: activate
```

Disabling one or both auto-lifecycle event switches will also remove the corresponding events from the YAML, while
enabling the switch will regenerate the event in the YAML.

## Setting component parameters

Component parameters can be edited directly in the settings. Try changing the value of the Timeout parameter to a

> different one and look for the corresponding change in the code editor. Then press play on the application and observe

Reviewer 3: just personal preference but I would remove "to a different one"

that the timer now has a different duration before triggering the "Is timed out" predicate.

## Adding and deleting elements

Let's now make the example a bit more interesting by adding more elements to the graph using the sidebar menu; press the
(+) button in the corner of the graph to open the sidebar. At the top, you will see special entries
"Hardware Interface", "Trigger Events Button", "Sequence" and "Condition", for which a dedicated example is included
in later sections. These are followed by a list of components with a name and brief description.

Trigger Events Buttons are interactive elements that can be used to trigger events through mouse clicks and interact
with the flow of the application. Click on the Trigger Events Button in the sidebar to add such an element to the
application. A new button should appear on the graph and in the code.

> The sidebar menu can be closed by clicking anywhere in the graph. Drag on the button to change its position in the

Reviewer 3: Personal pref: omit "on" in "drag on" line 66

Author: I think that's a leftover from a previous sentence yes

graph. This will also update the corresponding Trigger Events Button position in the code editor.

To delete an element from the graph, press the small menu icon in its top right corner, then click Remove.

## Creating and deleting event edges

> The timers are already connected with a Transition event edge. To change the event type on an edge, click on the label.

Reviewer 3: The sentence starting with "to change the event type on an edge..." is confusing to me, not sure if it is the usage of the word label or maybe the order of sentences between this and the next one where I feel its not quite clear why/ what we can change to.

Author: hmm, feels quite clear to me. Not sure how I could improve it

Author: [Screencast from 2024-12-19 11-18-40.webm](https://github.com/user-attachments/assets/ff8781d8-07ae-4a5f-8eb4-27c0047dd60a)

Talking about this, do you have a suggestion?

Reviewer 3: Sure, my point was moreso making it clear why the user would want to change the edge from transition to something else -- but maybe they already have that knowledge from elsewhere? @domire8
